routing/route.cpp

```diff
@@ -354,6 +359,42 @@ void Route::Update()
   m_currentTime = 0.0;
 }
 
+void Route::AppendTraffic(Route const & route)
+{
+  CHECK(route.IsValid(), ());
+
+  if (GetTraffic().empty() && route.GetTraffic().empty())
+    return;
+
+  if (!IsValid())
+  {
+    m_traffic.assign(route.GetTraffic().cbegin(), route.GetTraffic().cend());
+    return;
+  }
+
+  // Note. At this point the last item of |m_poly| should be removed.
+  // So the size of |m_traffic| should be equal to size of |m_poly.
+  if (GetTraffic().empty())
+  {
+    CHECK_GREATER_OR_EQUAL(m_poly.GetPolyline().GetSize(), 1, ());
+    m_traffic.resize(m_poly.GetPolyline().GetSize(), SpeedGroup::Unknown);
```

Comment:
> Why? Traffic is on segments, not on points, therefore the size of traffic should be m_poly.GetPolyline().GetSize() - 1, isn't it?

Please see my note above in the code AppendTraffic(). The thing is AppendTraffic() is called form Route::AppendRoute() and at this moment last point of m_poly.GetPolyline() has been removed in Route::AppendRoute().

> The method is called "Append", but here I see that you actually assign something to traffic.

Please see code below, after the line:
```c++
...
  if (route.GetTraffic().empty())
  {
    CHECK_GREATER_OR_EQUAL(route.m_poly.GetPolyline().GetSize(), 1, ());
    m_traffic.insert(m_traffic.end(),
                     route.m_poly.GetPolyline().GetSize() - 1 /* segment number is less by one */,
                     SpeedGroup::Unknown);
  }
  else
  {
    m_traffic.insert(m_traffic.end(), route.GetTraffic().cbegin(), route.GetTraffic().cend());
  }
```

> Is it OK that the m_traffic may not be empty before this line of code?

Ok, m_traffic could be empty (if there's no traffic information for this kind of route) or the size of |m_traffic| should be one less then |m_poly.GetPolyline().GetSize()|. There's a tricky when we need to append the route with empty |m_traffic| to the route with full |m_traffic|. It could happens in case of cross mwm car routing form A* route to osrm route. In that case the resulting route should have size |m_traffic| one less then |m_poly.GetPolyline().GetSize()|. In that case for segments which don't have any traffic information we set SpeedGroup::Unknown.